airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/StateManager.kt

```diff
@@ -91,48 +91,53 @@ class StateManager(
     fun checkpoint(): List<AirbyteStateMessage> =
         listOfNotNull(global?.checkpoint()) + nonGlobal.mapNotNull { it.value.checkpoint() }
 
+    private sealed interface StateForCheckpoint {
+        val opaqueStateValue: OpaqueStateValue?
+        val numRecords: Long
+    }
+
+    private data class Fresh(
+        override val opaqueStateValue: OpaqueStateValue,
+        override val numRecords: Long,
+    ) : StateForCheckpoint
+
+    private data class Stale(
+        override val opaqueStateValue: OpaqueStateValue?,
+    ) : StateForCheckpoint {
+        override val numRecords: Long
+            get() = 0L
+    }
+
     private sealed class BaseStateManager<K : Feed>(
         override val feed: K,
         initialState: OpaqueStateValue?,
-        private val isCheckpointUnique: Boolean = true,
     ) : StateManagerScopedToFeed {
-        private var current: OpaqueStateValue?
-        private var pending: OpaqueStateValue?
-        private var isPending: Boolean
-        private var pendingNumRecords: Long
-
-        init {
-            synchronized(this) {
-                current = initialState
-                pending = initialState
-                isPending = initialState != null
-                pendingNumRecords = 0L
-            }
-        }
+        private var currentStateValue: OpaqueStateValue? = initialState
+        private var pendingStateValue: OpaqueStateValue? = initialState
+        private var pendingNumRecords: Long = 0L
 
-        override fun current(): OpaqueStateValue? = synchronized(this) { current }
+        @Synchronized override fun current(): OpaqueStateValue? = currentStateValue
 
+        @Synchronized
         override fun set(
             state: OpaqueStateValue,
             numRecords: Long,
         ) {
-            synchronized(this) {
-                pending = state
-                isPending = true
-                pendingNumRecords += numRecords
-            }
+            pendingStateValue = state
+            pendingNumRecords += numRecords
         }
 
-        fun swap(): Pair<OpaqueStateValue?, Long>? {
-            synchronized(this) {
-                if (isCheckpointUnique && !isPending) {
-                    return null
+        @Synchronized
+        fun takeForCheckpoint(): StateForCheckpoint {
+            val stateForCheckpoint: StateForCheckpoint =
+                when (val pending = pendingStateValue) {
```

Comment:
nit: why put it in pending rather than `when (pendingStateValue)`?

Reply:
The kotlin compiler is unable to perform smart-casting. It's annoying. I've rewritten this expression differently.